File: Assignment_27/program2.c
// Write a program which accept string from user and copy the contents of that string into another string.
//  (Implement strncpy() function)

///////////////////////////////////////////////////////////////////////////////
// File Name   :  program2.c
// Description :  Accept string from user and  copy the contents of that string into another string. 
// Autor       :  Ankita Anil Patil
// Date        :  06/06/2025
///////////////////////////////////////////////////////////////////////////////



/*
    Input : "Marvellous Multi OS"
             10
    Output : "Marvellous 
*/

//Note : If third parameter is greater than the size of source string then copy whole string into destination.

#include<stdio.h>

void StrNCpyX(char *src, char *dest, int iCnt)
{
    if(src ==NULL)
    {
        printf("Invalid input ");
        return -1;
    }

    while((*src != '\0') && (iCnt != 0) )
    {
        *dest= *src;
         src++;
         dest++;
         iCnt--;
    }
    dest = '\0';
    }
int main()
{
    char arr[30] = "Marvellous Multi OS";
    char brr[30];
    
    StrNCpyX(arr,brr,10);
    
    printf("Copied string is : %s\n",brr);
   
    return 0;
}

File: Assignment_57/program2.java
// Given an array arr[] of n integers, 
// construct a Product Array  prod[] (of same size) such that prod[i] is equal to 
// the product of all the elements of arr[] except arr[i]. Solve it without division operator and in O(n).  

/*
    Input :  arr[] = {10, 3, 5, 6, 2} 
    Output : prod[] = {180, 600, 360, 300, 900}

 */

////////////////////////////////////////////////////////////////////////////////
// File Name   :  program2.java
// Description :  accept array arr[] of n integers, 
//                construct a Product Array  prod[] (of same size) such that 
//                prod[i] is equal to the product of all the elements of arr[]

//                except arr[i]. Solve it without division operator and in O(n).   
// Autor       :  Ankita Anil Patil
// Date        :  14/09/2025
///////////////////////////////////////////////////////////////////////////////

 class ProductArray 
 {
        public void productArray(int arr[], int n)
        {

            int[] left = new int[n];
            int[] right = new int[n];
            int[] prod = new int[n];

            
            left[0] = 1;
            for (int i = 1; i < n; i++) 
            {
                left[i] = arr[i - 1] * left[i - 1];
            }

        
            right[n - 1] = 1;
            for (int j = n - 2; j >= 0; j--) 
            {
                right[j] = arr[j + 1] * right[j + 1];
            }

        
            for (int i = 0; i < n; i++) 
            {
                prod[i] = left[i] * right[i];
            }

            for (int i = 0; i < n; i++)
            {
                System.out.print(prod[i] + "\t");
            }
    }
}

class program2 {
    public static void main(String[] args) {
        ProductArray pa = new ProductArray();
        int arr[] = {10, 3, 5, 6, 2};
        int n = arr.length;

        System.out.println("The product array is : ");
        pa.productArray(arr, n);
    }
}
